### PR TITLE
Fix the issue where the hash script property can't be reversed

### LINKED ISSUE #7422: Hash script property doesn't get reversed
**Describe the bug (REQUIRED)**
Trying to print a hash script property displays `hash: [XXXX (unknown)]`, even in debug builds.

**Expected behavior (REQUIRED)**
It should print the original value that was hashed.

**To Reproduce (REQUIRED)**
1. Create a script property like `go.property("my_hash", hash("test"))`
2. Try to print it

**Defold version (REQUIRED):**
  - 1.4.2

**Platforms (REQUIRED):**
 - Pop!_OS 22.04 LTS

**Minimal repro case (OPTIONAL):**
```lua
go.property("my_hash", hash("test"))

function init(self)
    print(self.my_hash) --> "hash: [12722210162474435462 (unknown)]"
end
```

**Workaround (OPTIONAL):**
Calling `hash()` with the same value at any point before printing the property will make it display the correct value:
```lua
go.property("my_hash", hash("test"))

function init(self)
    print(self.my_hash) --> "hash: [12722210162474435462 (unknown)]"
    hash("test")
    print(self.my_hash) --> "hash: [test]"
end
```

### PR DESCRIPTION
**Pull request by AGulev**
Hash values can be reversed to their string equivalents in Debug builds, simplifying the development process. This fix makes hashes added in `go.property()` reversible as well.

Fix https://github.com/defold/defold/issues/7422
## PR checklist

* [x] Code
	* [x] Add engine and/or editor unit tests.
	* [x] New and changed code follows the overall code style of existing code
	* [x] Add comments where needed
* [ ] Documentation
	* [ ] Make sure that API documentation is updated in code comments
	* [ ] Make sure that manuals are updated (in github.com/defold/doc)
* [x] Prepare pull request and affected issue for automatic release notes generator
	* [x] Pull request - Write a message that explains what this pull request does. What was the problem? How was it solved? What are the changes to APIs or the new APIs introduced? This message will be used in the generated release notes. Make sure it is well written and understandable for a user of Defold.
	* [x] Pull request - Write a pull request title that in a sentence summarises what the pull request does. Do not include "Issue-1234 ..." in the title. This text will be used in the generated release notes.
	* [x] Pull request - Link the pull request to the issue(s) it is closing. Use on of the [approved closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).
	* [x] Affected issue - Assign the issue to a project. Do not assign the pull request to a project if there is an issue which the pull request closes.
	* [ ] Affected issue - Assign the "breaking change" label to the issue if introducing a breaking change.
	* [ ] Affected issue - Assign the "skip release notes" is the issue should not be included in the generated release notes.

----------

Example of a well written PR description:

1. Start with the user facing changes. This will end up in the release notes.
1. Add one of the GitHub approved closing keywords
1. Optionally also add the technical changes made. This is information that might help the reviewer. It will not show up in the release notes. Technical changes are identified by a line starting with one of these:
   1. `### Technical changes` 
   1. `Technical changes:`
   2. `Technical notes:`

```
There was a anomaly in the carbon chroniton propeller, introduced in version 8.10.2. This fix will make sure to reset the phaser collector on application startup.

Fixes #1234

### Technical changes
* Pay special attention to line 23 of phaser_collector.clj as it contains some interesting optimizations
* The propeller code was not taking into account a negative phase.
```

### CODE REVIEW

File: com.dynamo.cr/com.dynamo.cr.bob.test/src/com/dynamo/bob/pipeline/LuaScannerTest.java [hunk 1]
```diff
@@ -507,4 +507,32 @@ public void testSemicolon() throws Exception {
         expected = "  ";
         assertEquals(expected, scanner.getParsedLua());
     }
+
+    @Test
+    public void testGoPropertyHash() throws Exception {
```
Comment (AGulev) [inline com.dynamo.cr/com.dynamo.cr.bob.test/src/com/dynamo/bob/pipeline/LuaScannerTest.java:512]: Tests for both debug and non-debug bundles

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/LuaBuilder.java [hunk 2]
```diff
@@ -108,6 +108,10 @@ private LuaScanner getLuaScanner(IResource resource) throws IOException, Compile
             }
 
             scanner = new LuaScanner();
+            if (variant == Bob.VARIANT_DEBUG)
+            {
+                scanner.setDebug();
```
Comment (AGulev) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/LuaBuilder.java:113]: We should keep hash only in debug builds

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/LuaScanner.java [hunk 3]
```diff
@@ -352,15 +358,18 @@ public void enterFunctioncall(LuaParser.FunctioncallContext ctx) {
...
+            paramsContext = argsCtx;
             String firstArg = getFirstStringArg(argsCtx);
             List<Token> tokens = getTokens(ctx, Token.DEFAULT_CHANNEL);
             Property property = new Property(tokens.get(0).getLine() - 1);
             property.name = firstArg;
             if (firstArg == null) {
                 property.status = Status.INVALID_ARGS;
             } else {
-                if (parsePropertyValue(argsCtx, property)) {
+                paramsContext = parsePropertyValue(argsCtx, property);
```
Comment (AGulev) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/LuaScanner.java:371]: To be able to keep hash tokens, we need better return value than just a bool, taking into account what this function does, it's fine to return `ParserRuleContext` and use it for both the error's checking and getting tokens we wanna keep

File: editor/src/clj/editor/code/script_compilation.clj [hunk 4]
```diff
@@ -147,31 +147,49 @@
...
+                              new-cursor-ranges (conj! cursor-ranges cursor-range-before-skipped)
+                              [_ start-row-after-skipped start-col-after-skipped] token
+                              start-cursor-after-skipped (data/->Cursor start-row-after-skipped start-col-after-skipped)
+                              cursor-range (data/->CursorRange start-cursor-after-skipped end-cursor)]
+                          (recur (conj! new-cursor-ranges cursor-range)
+                                 next-tokens
+                                 0
+                                 []
+                                 nil))))))
+            "hash" (recur cursor-ranges (next tokens) paren-count consumed token)
```
Comment (AGulev) [inline editor/src/clj/editor/code/script_compilation.clj:190]: new `skipped` token added into loop

File: editor/test/editor/code/script_test.clj [hunk 5]
```diff
@@ -245,7 +245,7 @@
     ["             "]
 
     ["go.property('name', hash())"]
-    ["                           "]
+    ["                    hash() "]
```
Comment (AGulev) [inline editor/test/editor/code/script_test.clj:248]: fix test

File: editor/src/clj/editor/code/script_compilation.clj [hunk 6]
```diff
@@ -147,31 +147,48 @@
...
-            (recur cursor-ranges (next tokens) paren-count consumed)))
+                          start-cursor (data/->Cursor start-row start-col)]
+                      (if (nil? skipped)
+                        (let [cursor-range (data/->CursorRange start-cursor end-cursor)]
+                          (recur (conj! cursor-ranges cursor-range)
+                                 next-tokens
+                                 0
+                                 []
+                                 nil))
+                        (let [[_ end-row-skipped end-col-skipped] skipped
```
Comment (AGulev) [inline editor/src/clj/editor/code/script_compilation.clj:177]: If we have a token that we want to skip, two cursor ranges should be produced: one for the "before skip" token and another for `)` or `);`.

**Review by britzl — APPROVED** [review]
The java part looks ok!

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/LuaScanner.java [hunk 7]
```diff
@@ -369,7 +378,17 @@ else if (fnDesc.is("property", "go")) {
             properties.add(property);
 
             // strip property from code
-            removeTokens(tokens, true);
+            if (isDebug) {
+              // keep tokens for hash() in debug build
+              // see https://github.com/defold/defold/issues/7422
```
Comment (matgis) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/LuaScanner.java:383]: Whitespace: Indentation looks off here?

**Review by matgis — APPROVED** [review]
LGTM!

**Pull request merged**